auth/credentials/idtoken/file.go

```diff
@@ -125,3 +112,77 @@ func credsFromBytes(b []byte, opts *Options) (*auth.Credentials, error) {
 		return nil, fmt.Errorf("idtoken: unsupported credentials type: %v", t)
 	}
 }
+
+func new2LOTokenProvider(f *credsfile.ServiceAccountFile, opts *Options) (auth.TokenProvider, error) {
+	opts2LO := &auth.Options2LO{
+		Email:        f.ClientEmail,
+		PrivateKey:   []byte(f.PrivateKey),
+		PrivateKeyID: f.PrivateKeyID,
+		TokenURL:     f.TokenURL,
+		UseIDToken:   true,
+	}
+	if opts2LO.TokenURL == "" {
+		opts2LO.TokenURL = jwtTokenURL
+	}
+
+	var customClaims map[string]interface{}
+	if opts != nil {
+		customClaims = opts.CustomClaims
+	}
+	if customClaims == nil {
+		customClaims = make(map[string]interface{})
+	}
+	customClaims["target_audience"] = opts.Audience
+
+	opts2LO.PrivateClaims = customClaims
+	return auth.New2LOTokenProvider(opts2LO)
+}
+
+// newIAMIDTokenProvider creates a TokenProvider that performs an authenticated
+// RPC with the IAM service to obtain an ID token. The provided service account
+// must have the iam.serviceAccountTokenCreator role. If a fully-authenticated
+// client is not provided, the service account must support a self-signed JWT.
+// This TokenProvider is primarily intended for use in non-GDU universes, which
+// do not have access to the oauth2.googleapis.com/token endpoint, and thus must
+// use IAM generateIdToken instead.
+func newIAMIDTokenProvider(b []byte, f *credsfile.ServiceAccountFile, opts *Options) (auth.TokenProvider, error) {
+	var client *http.Client
+	var creds *auth.Credentials
+	var err error
+	if opts.Client == nil {
+		creds, err = credentials.DetectDefault(&credentials.DetectOptions{
+			CredentialsJSON:  b,
+			Scopes:           []string{"https://www.googleapis.com/auth/iam"},
+			UseSelfSignedJWT: true,
+		})
+		if err != nil {
+			return nil, err
+		}
+		client, err = httptransport.NewClient(&httptransport.Options{
+			Credentials:    creds,
+			UniverseDomain: opts.UniverseDomain,
+		})
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		client = opts.Client
+	}
+	its := iamIDTokenProvider{
```

Comment:
Is there a reason not to just call the impersonate package instead of duplicating this code? Or could they share code in a local internal?

Reply:
I think you're right, I'll try again to refactor out duplicated code, thanks.